### PR TITLE
ARROW-10487 [FlightRPC][C++] Header-based auth in clients

### PR DESCRIPTION
**Pull request by lyndonbauto**
Added support for header based authentication in clients.
- Added support for base 64 encoded username / password auth to match Java implementation
- Added bearer token receiving and populating of call options to send back
- Added unit tests that connects C++ client and a mock C++ server to test authentication

### CODE REVIEW

**Comment by lyndonb-bq** [comment]
@lidavidm 
Hi David,
This is currently a work in progress. I was wondering if you could take a look at how this was integrated into the C++ library and let me know if it is okay or if there is a better way to do it? Afterwards I will make any required code changes and fix the code linting/commenting.
Thanks!

File: cpp/src/arrow/flight/client.cc [hunk 1]
```diff
@@ -51,6 +51,7 @@
 
 #include "arrow/flight/client_auth.h"
 #include "arrow/flight/client_middleware.h"
+#include "arrow/flight/client_header_auth_middleware.h"
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client.cc:54]: Is this ordering alphabetical? If so this should be up one...

File: cpp/src/arrow/flight/client.cc [hunk 2]
```diff
@@ -328,7 +332,7 @@ class GrpcClientInterceptorAdapterFactory
     : public grpc::experimental::ClientInterceptorFactoryInterface {
  public:
   GrpcClientInterceptorAdapterFactory(
-      std::vector<std::shared_ptr<ClientMiddlewareFactory>> middleware)
+      std::vector<std::shared_ptr<ClientMiddlewareFactory>>& middleware)
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client.cc:335]: Should this be const?

File: cpp/src/arrow/flight/client.cc [hunk 3]
```diff
@@ -371,7 +375,7 @@ class GrpcClientInterceptorAdapterFactory
   }
 
  private:
-  std::vector<std::shared_ptr<ClientMiddlewareFactory>> middleware_;
+  std::vector<std::shared_ptr<ClientMiddlewareFactory>>& middleware_;
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client.cc:378]: Intention is to keep the reference and not a copy?

File: cpp/src/arrow/flight/client.cc [hunk 4]
```diff
@@ -963,8 +967,9 @@ class FlightClient::FlightClientImpl {
 
     std::vector<std::unique_ptr<grpc::experimental::ClientInterceptorFactoryInterface>>
         interceptors;
+    middleware = std::move(options.middleware);
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client.cc:970]: Why this change?

File: cpp/src/arrow/flight/client.cc [hunk 5]
```diff
@@ -993,6 +998,30 @@ class FlightClient::FlightClientImpl {
     return Status::OK();
   }
 
+  Status AuthenticateBasicToken(std::string username, std::string password,
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client.cc:1001]: Should username/password be passed by const ref?

File: cpp/src/arrow/flight/client.cc [hunk 6]
```diff
@@ -993,6 +998,30 @@ class FlightClient::FlightClientImpl {
...
+                                std::pair<std::string, std::string>* bearer_token) {
+    // Add bearer token factory to middleware so it can intercept the bearer token.
+    middleware.push_back(std::make_shared<ClientBearerTokenFactory>(bearer_token));
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client.cc:1004]: This looks odd to create the shared pointer after you've passed in the raw pointer....it seems like the method itself should take a shared pointer.

File: cpp/src/arrow/flight/client.cc (hunk 1 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.cc:54]: Yes, was waiting for feedback on design before proper formatting. I will correct this.

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  ClientBearerTokenMiddleware::ClientBearerTokenMiddleware(
+    std::pair<std::string, std::string>* bearer_token_)
+        : bearer_token(bearer_token_) { }
+
+  void ClientBearerTokenMiddleware::SendingHeaders(AddCallHeaders* outgoing_headers) { }
+
+  void ClientBearerTokenMiddleware::ReceivedHeaders(
+    const CallHeaders& incoming_headers) {
+    // Grab the auth token if one exists.
+    auto bearer_iter = incoming_headers.find(AUTH_HEADER);
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:40]: const?

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    if (bearer_iter == incoming_headers.end()) {
+      return;
+    }
+
+    // Check if the value of the auth token starts with the bearer prefix, latch the token.
+    std::string bearer_val = bearer_iter->second.to_string();
+    if (bearer_val.size() > BEARER_PREFIX.size()) {
+      bool hasPrefix = std::equal(bearer_val.begin(), bearer_val.begin() + BEARER_PREFIX.size(), BEARER_PREFIX.begin(),
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:48]: Seems like they use snake_case instead of camelCase.

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      if (hasPrefix) {
+        *bearer_token = std::make_pair(AUTH_HEADER, bearer_val);
+      }
+    }
+  }
+
+  void ClientBearerTokenMiddleware::CallCompleted(const Status& status) { }
+
+  void ClientBearerTokenFactory::StartCall(const CallInfo& info, std::unique_ptr<ClientMiddleware>* middleware) {
+    *middleware = std::unique_ptr<ClientBearerTokenMiddleware>(new ClientBearerTokenMiddleware(bearer_token));
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:62]: std::make_unique?

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      );
+      if (hasPrefix) {
+        *bearer_token = std::make_pair(AUTH_HEADER, bearer_val);
+      }
+    }
+  }
+
+  void ClientBearerTokenMiddleware::CallCompleted(const Status& status) { }
+
+  void ClientBearerTokenFactory::StartCall(const CallInfo& info, std::unique_ptr<ClientMiddleware>* middleware) {
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:61]: Would it be better to pass a reference instead of a pointer?

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+  void ClientBearerTokenFactory::Reset() {
+    *bearer_token = std::make_pair("", "");
+  }
+
+  template<typename ... Args>
+  std::string string_format(const std::string& format, const Args... args) {
+    // Check size requirement for new string and increment by 1 for null terminator.
+    size_t size = std::snprintf(nullptr, 0, format.c_str(), args ...) + 1;
+    if(size <= 0){
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:73]: Nit: spacing between if (, ){

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    return std::string(buf.get(), buf.get() + size - 1);
+  }
+
+  void AddBasicAuthHeaders(grpc::ClientContext* context, const std::string& username, const std::string& password) {
+    const std::string formatted_credentials = string_format("%s:%s", username.c_str(), password.c_str());
+    context->AddMetadata(AUTH_HEADER, BASIC_PREFIX + base64_encode(formatted_credentials));
+  }
+
+  std::string base64_encode(const std::string& input) {
+     static const std::string base64_chars =
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:91]: This doesn't exist in the codebase already?

**Comment by lidavidm** [comment]
Note, for formatting, use clang-format; for naming conventions, please follow Google C++ guidelines. See the [developer docs](https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci).

File: cpp/src/arrow/flight/client_header_auth_middleware.h [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+#else
+#include <grpc++/grpc++.h>
+#endif
+
+#include <algorithm>
+#include <iostream>
+#include <cctype>
+#include <string>
+
+const std::string AUTH_HEADER = "authorization";
```
Comment (kylepbit) [inline cpp/src/arrow/flight/client_header_auth_middleware.h:41]: Do these need to be in the header, or should they be only in the CC.
If in the header, I think they're created anew for each compilation unit, and thus should be defined as extern with the actual value defined in the CC to avoid multiple instantiations.

File: cpp/src/arrow/flight/client.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/client.cc:335]: This should either be const reference or a pointer.

File: cpp/src/arrow/flight/client.cc (hunk 3 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/client.cc:378]: This should be a pointer if not a copy.

File: cpp/src/arrow/flight/client.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/client.cc:335]: So, I'd rather we store a reference to this interceptor factory instead of a reference to the middleware with a mutable pointer back, i.e. I'd rather have this class own the middleware as it currently does, and have FlightClient call a method of this class to add more middleware at runtime.

File: cpp/src/arrow/flight/client.h [hunk 14]
```diff
@@ -65,6 +65,9 @@ class ARROW_FLIGHT_EXPORT FlightCallOptions {
 
   /// \brief IPC writer options, if applicable for the call.
   ipc::IpcWriteOptions write_options;
+
+  /// \brief Metadata for client to add to context.
+  std::vector<std::pair<std::string, std::string>> metadata;
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.h:70]: I believe we call it headers elsewhere, so this should stay consistent with that.

File: cpp/src/arrow/flight/client.h [hunk 15]
```diff
@@ -191,6 +194,14 @@ class ARROW_FLIGHT_EXPORT FlightClient {
   Status Authenticate(const FlightCallOptions& options,
                       std::unique_ptr<ClientAuthHandler> auth_handler);
 
+  /// \brief Authenticate to the server using the given handler.
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.h:197]: There's no handler in play here.

File: cpp/src/arrow/flight/client_header_auth_middleware.h [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+#pragma once
+
+#include "arrow/flight/client_middleware.h"
+#include "arrow/flight/client_auth.h"
+#include "arrow/flight/client.h"
+
+#ifdef GRPCPP_PP_INCLUDE
+#include <grpcpp/grpcpp.h>
+#if defined(GRPC_NAMESPACE_FOR_TLS_CREDENTIALS_OPTIONS)
+#include <grpcpp/security/tls_credentials_options.h>
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_auth_middleware.h:30]: I don't think we need this include?

File: cpp/src/arrow/flight/client_header_auth_middleware.h (hunk 13 above)
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_auth_middleware.h:41]: These should be named kAuthHeader etc. and should go in the .cc file unless we actually want to expose these to users.

Or alternatively, the other constants are currently in one of the _internal.h headers.

File: cpp/src/arrow/flight/client_header_auth_middleware.h [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+const std::string AUTH_HEADER = "authorization";
+const std::string BEARER_PREFIX = "Bearer ";
+const std::string BASIC_PREFIX = "Basic ";
+
+namespace arrow {
+namespace flight {
+
+// TODO: Need to add documentation in this file.
+void ARROW_FLIGHT_EXPORT AddBasicAuthHeaders(grpc::ClientContext* context,
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_auth_middleware.h:49]: This is internal - it shouldn't be in a public header. (Ditto for the grpc include.)

File: cpp/src/arrow/flight/client_header_auth_middleware.h [hunk 18]
```diff
@@ -0,0 +1,78 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_auth_middleware.h:1]: This should probably be an _internal.h header since I don't think any of this is intended to be directly used outside of the implementation here.

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+  void ClientBearerTokenFactory::StartCall(const CallInfo& info, std::unique_ptr<ClientMiddleware>* middleware) {
+    *middleware = std::unique_ptr<ClientBearerTokenMiddleware>(new ClientBearerTokenMiddleware(bearer_token));
+  }
+
+  void ClientBearerTokenFactory::Reset() {
+    *bearer_token = std::make_pair("", "");
+  }
+
+  template<typename ... Args>
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:69]: I don't think we need an entire templated function to concatenate two strings.

File: cpp/src/arrow/flight/client_header_auth_middleware.cc [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+  std::string string_format(const std::string& format, const Args... args) {
+    // Check size requirement for new string and increment by 1 for null terminator.
+    size_t size = std::snprintf(nullptr, 0, format.c_str(), args ...) + 1;
+    if(size <= 0){
+      throw std::runtime_error("Error during string formatting. Format: '" + format + "'.");
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_auth_middleware.cc:74]: And Arrow disallows exceptions.

File: cpp/src/arrow/flight/client.cc (hunk 3 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.cc:378]: It was to keep the reference, but you got me thinking more and I think I should just take a copy and then make a function to allow adding to the middleware on the fly, since the issue is that I need to be able to add and remove middleware from the interceptor. I will rejig this.

File: cpp/src/arrow/flight/client.cc (hunk 4 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.cc:970]: Goal was so that middleware was retained in this class so I could push and pop it. I am going to adjust the implementation though.

File: cpp/src/arrow/flight/client.cc (hunk 5 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.cc:1001]: yes

**Review by lidavidm — CHANGES_REQUESTED** [review]
Some initial comments.

Also:
- We should have unit tests, not just integration tests
- If you want this to actually run as part of the integration suite, you'll have to add it to the list of [test scenarios](https://github.com/apache/arrow/blob/master/dev/archery/archery/integration/runner.py#L344) and various other places; maybe take a look at the commit:
https://github.com/apache/arrow/commit/0188e45bbe688c45f10032e9c37fbedab755fa71

File: cpp/src/arrow/flight/client.cc (hunk 6 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.cc:1004]: The raw pointer is unpopulated, so it's passed to the BearerTokenFactory's constructor, which stores it and populated it when it receives the bearer token. I could make the client pass the whole factory in with the bearer token already inside it, but it's more work and requires they understand what's going on more than they otherwise would need to.

File: cpp/src/arrow/flight/client.h (hunk 14 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.h:70]: Will correct this.

File: cpp/src/arrow/flight/client_header_auth_middleware.cc (hunk 12 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client_header_auth_middleware.cc:91]: Yes, I found it. Will remove this.

File: cpp/src/arrow/flight/client_header_auth_middleware.cc (hunk 10 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client_header_auth_middleware.cc:61]: Can't this is a method of the base class, I think it's done this way to allow you to assign a new unique pointer to it without exposing the other middlewares they are already holding in their vector.

File: cpp/src/arrow/flight/client.cc [hunk 21]
```diff
@@ -1175,6 +1222,7 @@ class FlightClient::FlightClientImpl {
       noop_auth_check_;
 #endif
   int64_t write_size_limit_bytes_;
+  GrpcClientInterceptorAdapterFactory* interceptor_pointer;
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.cc:1225]: Please follow naming conventions (`interceptor_pointer_`)

File: cpp/src/arrow/flight/client.cc [hunk 22]
```diff
@@ -993,6 +1010,36 @@ class FlightClient::FlightClientImpl {
...
+                                std::pair<std::string, std::string>* bearer_token) {
+    // Add bearer token factory to middleware so it can intercept the bearer token.
+    if (interceptor_pointer != NULLPTR) {
+      interceptor_pointer->AddMiddlewareFactory(
+          std::make_shared<internal::ClientBearerTokenFactory>(bearer_token));
+    } else {
+      return MakeFlightError(FlightStatusCode::Internal,
+                             "Connect must be called before AuthenticateBasicToken.");
+    }
+    ClientRpc rpc({});
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.cc:1023]: You may actually want to allow passing call options so things like timeouts can be set.

File: cpp/src/arrow/flight/client.cc [hunk 23]
```diff
@@ -993,6 +1010,36 @@ class FlightClient::FlightClientImpl {
     return Status::OK();
   }
 
+  Status AuthenticateBasicToken(const std::string& username, const std::string& password,
+                                std::pair<std::string, std::string>* bearer_token) {
+    // Add bearer token factory to middleware so it can intercept the bearer token.
+    if (interceptor_pointer != NULLPTR) {
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.cc:1016]: Honestly, instead of using interceptors, why not use ClientRpc.context (a [grpc::ClientContext](https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html)) to attach the headers (AddMetadata) and retrieve them (GetServerInitialMetadata)? This isn't a general auth mechanism anyways, so I'd rather we keep the implementation simple & tightly scoped in this case.

File: cpp/src/arrow/flight/client.h [hunk 24]
```diff
@@ -191,6 +194,14 @@ class ARROW_FLIGHT_EXPORT FlightClient {
   Status Authenticate(const FlightCallOptions& options,
                       std::unique_ptr<ClientAuthHandler> auth_handler);
 
+  /// \brief Authenticate to the server using basic authentication with base 64 encoding.
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.h:197]: nit: base64 encoding is part of the definition of HTTP basic auth

**Comment by lidavidm** [comment]
By the way, for CMake formatting, please use [cmake_format](https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci)

File: cpp/src/arrow/flight/client.cc [hunk 25]
```diff
@@ -1198,6 +1227,12 @@ Status FlightClient::Authenticate(const FlightCallOptions& options,
   return impl_->Authenticate(options, std::move(auth_handler));
 }
 
+Status FlightClient::AuthenticateBasicToken(
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.cc:1230]: Could we make this `arrow::Result<std::pair<>>`?

File: cpp/src/arrow/flight/flight_test.cc [hunk 26]
```diff
@@ -1010,6 +1140,57 @@ class TestErrorMiddleware : public ::testing::Test {
...
+    call_options.headers.push_back(bearer_token);
+    ASSERT_OK(client_->ListFlights(call_options, {}, &listing));
+    ASSERT_TRUE(bearer_middleware_->GetIsValid());
+  }
+
+  void RunInvalidClientAuth() {
+    std::pair<std::string, std::string> bearer_token;
+    // Note: Status intentionally ignored because it requires C++ server implementation of
+    // header auth. For now it returns an IOError.
+    arrow::Status status = client_->AuthenticateBasicToken(
```
Comment (lidavidm) [inline cpp/src/arrow/flight/flight_test.cc:1179]: If the intent is to test that this fails, we can `ASSERT_RAISES(IOError, status)` - even if we had an implementation of header auth, we'd want this to fail, right?

**Review by lidavidm — COMMENTED** [review]
LGTM - just two minor comments (one about the interface, one about a test)

File: cpp/src/arrow/flight/flight_test.cc (hunk 26 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/flight_test.cc:1179]: Sorry, this was an artifact of my initial test implementation which was incorrect on the test server, and always caused an error. I forgot to remove it from the negative test once I fixed it. I will correct this.

File: cpp/src/arrow/flight/client.cc (hunk 25 above)
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client.cc:1230]: Yep, I will make this change.

**Comment by lyndonb-bq** [comment]
@lidavidm Hey David, I have made the requested changes. I got a random Javascript failure in one of the CI teardowns, not sure if this is accepted or if you can restart the CI build (I do not have permission to restart it myself).

**Comment by lyndonb-bq** [comment]
> @lidavidm Hey David, I have made the requested changes. I got a random Javascript failure in one of the CI teardowns, not sure if this is accepted or if you can restart the CI build (I do not have permission to restart it myself).

@lidavidm All tests are now passing 🎉

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Thanks @lyndonb-bq!
